service-accounts.tf

```diff
@@ -1,88 +1,55 @@
-  project = var.project
-  role    = "roles/iam.serviceAccountUser"
-  member  = "serviceAccount:${google_service_account.default.email}"
-}
+  for_each = local.service_accounts
 
-resource "google_project_iam_member" "kube_system_service_account" {
   project = var.project
   role    = "roles/iam.serviceAccountUser"
-  member  = "serviceAccount:${google_service_account.kube_system.email}"
+  member  = "serviceAccount:${google_service_account.accounts[each.value].email}"
 }
 
-resource "google_project_iam_member" "kube_public_service_account" {
-  project = var.project
-  role    = "roles/iam.serviceAccountUser"
-  member  = "serviceAccount:${google_service_account.kube_public.email}"
-}
-
-resource "google_project_iam_member" "compute_service_account" {
-  project = var.project
-  role    = "roles/iam.serviceAccountUser"
-  member  = "serviceAccount:${google_service_account.compute.email}"
-}
-
-resource "google_project_iam_member" "platform_service_account" {
-  project = var.project
-  role    = "roles/iam.serviceAccountUser"
-  member  = "serviceAccount:${google_service_account.platform.email}"
+resource "google_project_iam_member" "platform_roles" {
+  for_each = toset(["roles/storage.objectAdmin", "roles/logging.logWriter", "roles/monitoring.metricWriter"])
+  project  = var.project
+  role     = each.value
+  member   = "serviceAccount:${google_service_account.accounts["platform"].email}"
 }
 
-resource "google_project_iam_member" "platform_object_admin" {
-  project = var.project
-  role    = "roles/storage.objectAdmin"
-  member  = "serviceAccount:${google_service_account.platform.email}"
+resource "google_service_account_iam_binding" "platform_gcs" {
+  service_account_id = google_service_account.accounts["platform"].name
+  role               = "roles/iam.workloadIdentityUser"
+  members = [
+    "serviceAccount:${var.project}.svc.id.goog[${var.platform_namespace}/docker-registry]",
+    "serviceAccount:${var.project}.svc.id.goog[${var.platform_namespace}/git]",
+    "serviceAccount:${var.project}.svc.id.goog[${var.platform_namespace}/nucleus]",
+  ]
 }
 
-resource "google_project_iam_member" "platform_workload_identity_user" {
-  project = var.project
-  role    = "roles/iam.workloadIdentityUser"
-  member  = "serviceAccount:${google_service_account.platform.email}"
-}
+resource "google_artifact_registry_repository_iam_member" "gcr" {
+  for_each = google_artifact_registry_repository.domino
+  provider = google-beta
 
-resource "google_project_iam_member" "platform_logging" {
-  project = var.project
-  role    = "roles/logging.logWriter"
-  member  = "serviceAccount:${google_service_account.platform.email}"
-}
+  repository = each.value.name
+  location   = each.value.location
 
-resource "google_project_iam_member" "platform_monitoring" {
-  project = var.project
-  role    = "roles/monitoring.metricWriter"
-  member  = "serviceAccount:${google_service_account.platform.email}"
+  role   = "roles/artifactregistry.writer"
```

Comment:
Is the read (pull) permission for the cluster implied or come from the object read?

Reply:
I know for GCR you get access within the project. I am waiting to see what happens to determine if I need to grant read for kubelets.